### ISSUE TITLE
Can't connect to iOS App

### ISSUE THREAD

**Issue by cipi1965** [samcm/PlexTogether#26]
In console it says 403 (Forbidden)

**Comment by samcm** [comment]
Nearly entirely certain the is similar to #19 as it appears the iOS app and ATV app use the same code base. I've spent a couple hours trying to get them working but haven't got any progress. It was working previously and none of the code on this side has changed so I'm half tempted to say an update to the apps broke the compatibility.

**Comment by cipi1965** [comment]
Did you read [this](https://github.com/plexinc/plex-media-player/wiki/Remote-control-API#general-guidelines)? It says all requests to Plex Players must include X-Plex-Target-Client-Identifier header

**Comment by samcm** [comment]
Yup all of that is implemented

**Comment by maciekish** [comment]
Any news on this please? Getting the same `403` :(

**Comment by pureMidi** [comment]
We haven't had a chance to look at this any further.

**Comment by maciekish** [comment]
Any pointers as to where to start? `assets/plex` perhaps? Im not a js developer but i might have a crack at it later in the day, hoping its something simple.

**Comment by cipi1965** [comment]
@maciekish I think you could start finding what is wrong with PlexTogether request to iOS app (Which you can simply analyze with some DevTools)

**Comment by pureMidi** [comment]
`/src/components/` is where the majority of what you'd be looking for is. As per the above, inspecting the traffic as it passes from PT to the client may be a good starting point too

**Comment by maciekish** [comment]
Looks like the problem is in fact `X-Plex-Target-Client-Identifier` missing. If i look at the traffic in Chrome the initial OPTIONS request to the player has no such header. If i do a regular GET request with the same headers via Postman i get a proper error saying `[kRemoteControlRouterErrorDomain] Header 'X-Plex-Target-Client-Identifier' not found (1)`

Maybe `X-Plex-Target-Client-Identifier` is implemented for the other but this one got omitted somehow?

**Comment by maciekish** [comment]
Seems like these three are the minimum to make the OPTIONS request go through `X-Plex-Target-Client-Identifier`, `X-Plex-Client-Identifier`, `X-Plex-Device-Name`.

**Comment by maciekish** [comment]
Furthermore it looks like `X-Plex-Target-Client-Identifier` etc is passed to request.js but they are not there in the request as shown by Chrome.

**Comment by pureMidi** [comment]
You're a champion. Thank you for going through the trouble. Might explain why some other clients are having some trouble. I'll make sure Sam sees this and takes a look when he's back from his work trip.

**Comment by maciekish** [comment]
No worries, i was hoping to fix it but it looks like more knowledge about Plex and js is required. Could `OPTIONS`/`GET` have something to do with the fact that the Windows client was local and the iOS client on another ip?

**Comment by samcm** [comment]
Hey @maciekish, thanks for having a look mate! I've just got back home so should have some time in the next couple days to have a look at this again.

**Comment by maciekish** [comment]
@samcm Sounds great, let me know if i can be of any more help!

**Comment by samcm** [comment]
I'm fairly sure that the Plex iOS/ATV apps not responding correctly to `OPTIONS` is the cause of this. As such, there isn't a whole lot that I can do from this side. Sending the exact same requests to other clients (that are on a different IP/Host) respond correctly to `OPTIONS`.

I originally had both of them working and haven't really changed any of the client interaction stuff since. Some people have had it work for them since 1.0, which could potentially be explained by them being on an old app version. 

I may spend some time over the weekend going through the app logs Regardless, there is a nasty bug with all of the iOS clients where the app dies after 5-10 minutes of timeline updates.

**Comment by maciekish** [comment]
Thanks for the update, but how do you explain the OPTIONS request not containing the proper headers? Are they not required there?

**Comment by maciekish** [comment]
I just tried something **_wild_**, i ran Plex on my iPad Pro and swiped in from the right to access the browser. Now i had both Plex and Plextogether running simultaneously on the "same ip". It connected just fine.

So the Windows client works when plextogether is open in a browser on the same machine, and the iOS client works if the browser is also open on the same device. It is the same iOS app version on both my phone and iPad, phone doesn't work when trying to connect from Windows browser but iPad works from local Safari.

**Could the problem simply be CORS?**

**Comment by samcm** [comment]
Yeah the problem at its core is CORS. I'm not certain on how it works but I think browsers send the Pre-flight `OPTIONS` request if the request is non-basic (the additional `X-Plex-*` headers mean this is the case) and the response must be valid for the following `GET`/`POST` to be allowed. The apps respond with `403` so it blocks further connections. 

The `X-Plex-*` headers are actually present in the pre-flight `OPTIONS` request if memory serves me right. They aren't in there in the standard form though, rather as a comma-separated list within some header. I think this is for the server to check if its OK to allow CORS depending on the content of this value. I think that is where this problem is arising - Plex isn't handling `X-Plex-*` headers via `OPTIONS` and just throws a `403`. 

Then again, I'm really not that familiar with how it works so I may be completely wrong. I've spent a large amount of time making the Plex Together Player mobile friendly and it works so that would the a work around when it's released.

**Comment by edgariscoding** [comment]
Any news on this? Everyone I know that I'd like to use this with also has an Apple TV with a MacBook Pro running Safari and gets the same error.

**Comment by samcm** [comment]
Nothing new to report sorry. Still looks like it's on Plex's side.

**Comment by brendenbus** [comment]
I am not sure if the issue has changed since it was originally reported but now when I try to connect to my AppleTV as the player I receive "Failed to load resource: the request timed out"

Not sure where it is retrieving 10.0.1.21 or 192.168.1.100 (which I assume is referring to my local network?) because my IP range is 192.168.22.x.

![screen shot 2017-09-01 at 4 48 40 pm](https://user-images.githubusercontent.com/823523/29987325-a329e1c6-8f35-11e7-8dcd-72a7bca2cd4a.png)

**Comment by samcm** [comment]
This issue should be fixed by the Plex team soon:

https://forums.plex.tv/discussion/304118/fix-synclounge-for-apple-tv

**Comment by samcm** [comment]
As per the above thread, AppleTV connectivity should be fixed now. Still waiting on iOS but shouldn't be too far behind. 

Side note: I'd strongly recommend increasing the `Plex Client Polling Interval` setting within SL Settings when using these clients. They REALLY don't like being constantly pinged.

**Comment by KyeRussell** [comment]
Dose this mean that this issue can be closed (and the README file updated to reflect that Apple TV is working?)

**Comment by SXRWahrheit** [comment]
Hi! Some friends (not me, I do not own one) are having issues with using their Apple TV clients and we're trying to do a watch party tonight. Any idea what could be wrong? Happy to get more info from them, also going to link them directly to this issue!

**Comment by samcm** [comment]
@SXRWahrheit What version of the Plex app are you running? You may have to manually update the app in the app store to get a up-to-date version that has the above fixes.
